Add parallax effect.

sparshg opened the pull request:
@chimosky @srevinsaju Please review, I splitted the background into different images and moved them with different speeds to make parallax effect.

sparshg committed:
Add parallax effect.

srevinsaju commented on welcomescreen.py:
```diff
@@ -34,6 +34,9 @@ class welcomescreen:
 
     def __init__(self, display):
         self.gameDisplay = display
+        self.x1_x2 = [[350, 350 + 490] for _ in range(4)]
```
How did you find these numbers, are they by trial and error? Also, do you think if moving these numbers to variables might help?

sparshg replied:
Unfortunately so many numbers are hardcoded everywhere already in the activity, 490 being the width of actual game. These numbers were already there in the form of `land1x = 350` and `land2x = 840` which I just converted in form of array.

chimosky replied:
Could you make these numbers constants so it'll be easier to know which is which and for future references? Thanks!

sparshg replied:
I extracted hardcoded values and also added more comments.

srevinsaju commented on welcomescreen.py:
```diff
@@ -77,16 +91,20 @@ def run(self):
-                land2x = 837
+            self.gameDisplay.blit(land1, (self.x1_x2[0][0], self.y[0]))
+            self.gameDisplay.blit(land2, (self.x1_x2[0][1], self.y[0]))
+            for i, speed in enumerate(self.x_speeds):
```
Can we have an inline comment explaining what happens in the following block?

sparshg replied:
Added comments.

chimosky commented on welcomescreen.py:
```diff
@@ -44,14 +47,23 @@ def run(self):
         land2 = land1
-        sky = pygame.image.load("assets/sky.png").convert()
-        sky = pygame.transform.scale(sky, (490, 200))
+        background = [
+            (i, i) for i in (
+                pygame.transform.scale(
+                    pygame.image.load("assets/sky3.png").convert(), (490, 33 * (200 / 109)),
+                ),
+                pygame.transform.scale(
+                    pygame.image.load("assets/sky2.png").convert_alpha(), (490, 40 * (200 / 109)),
+                ),
+                pygame.transform.scale(
+                    pygame.image.load("assets/sky1.png").convert_alpha(), (490, 44 * (200 / 109)),
+                ),
+            )
+        ]
```
This is quite difficult to read, could you make it better? Since the only thing that changes are the images then maybe a function that does that would be helpful.

chimosky commented:
Reviewed, not tested. Thanks!

sparshg committed:
extract hard coded values into variables.

sparshg committed:
Cleanup code structure.

Break hardcoded values into variables.
Cast `float` to `int` to support older versions of pygame.

sparshg commented:
I added more comments and broke more of hardcoded values into well defined variables. Also splitted the `self.background` list logic into more readable code.

chimosky commented on welcomescreen.py:
```diff
@@ -45,23 +52,23 @@ def run(self):
             (i, i) for i in (
                 pygame.transform.scale(
-                    pygame.image.load("assets/sky3.png").convert(), (490, 33 * (200 / 109)),
+                    pygame.image.load("assets/sky3.png").convert(), (self.width, 33 * (200 / 109)),
```
What about the other values?

sparshg replied:
Um.. I already changed that in latest commit, that is 81203462430a303073cd952961fc7b96362e6d30

chimosky replied:
Seen, thanks!

chimosky commented on main.py:
```diff
@@ -85,6 +84,32 @@ def initialize(self):
+        background = [
+            pygame.image.load("assets/sky3.png").convert(),
+            pygame.image.load("assets/sky2.png").convert(),
+            pygame.image.load("assets/sky1.png").convert_alpha()
```
I'm curious, why `convert_alpha` here?

sparshg replied:
sky1.png is overlapping sky2.png
Without convert_alpha, it won't take transparency into account

chimosky replied:
Okay, thanks!

chimosky commented:
Tested and I get this error;
```
Traceback (most recent call last):
  File "/home/ibiam/Activities/flappy-birds-activity/main.py", line 121, in make
    self.initialize()
  File "/home/ibiam/Activities/flappy-birds-activity/main.py", line 100, in initialize
    ), background)
  File "/home/ibiam/Activities/flappy-birds-activity/main.py", line 99, in <lambda>
    img, (self.width, img.get_height() * self.scaling_factor)
TypeError: integer argument expected, got float
1657153947.678694 ERROR root: Event queue full!
1657153947.693917 ERROR root: Event queue full!
1657153947.710997 ERROR root: Event queue full!
1657153947.728417 ERROR root: Event queue full!
1657153947.745286 ERROR root: Event queue full!
1657153947.761070 ERROR root: Event queue full!
1657153947.778083 ERROR root: Event queue full!
1657153947.793622 ERROR root: Event queue full!
1657153947.810643 ERROR root: Event queue full!
1657153947.827930 ERROR root: Event queue full!
```

sparshg commented:
Fixed, this didn't happen on my setup though, I think sugar uses older version of pygame. I fixed it on python 3.6 and pygame 1.9.4 and now it works on that.

chimosky commented:
Tested, works as expected.

> Fixed, this didn't happen on my setup though, I think sugar uses older version of pygame. I fixed it on python 3.6 and pygame 1.9.4 and now it works on that.

My tests were with pygame `v1.9.4`, could you also state the int conversion done in your commit message as that's part of the reasons why you made the commit.

chimosky commented:
Could you break the commit message to a title and a body that's after a new line? That'll be more helpful.
